pyccel/ast/headers.py

```diff
@@ -247,20 +252,23 @@ def build_argument(var_name, dc):
                         is_argument=True)
             return var
 
-        def process_template(arg_code, Tname, d_type, type_index):
-            tmp_arg_code = arg_code.copy()
-            resolved = False
+        def process_template(signature, Tname, d_type):
+            new_sig = tuple(d_type if 'datatype' in t and t['datatype'] == Tname\
+                    else t for t in signature)
+            return new_sig
 
-            for i, t in enumerate(tmp_arg_code):
-                if 'datatype' in t and t['datatype'] == Tname:
-                    resolved = True
-                    tmp_arg_code[i] = d_type.copy()
-            if resolved or type_index == 0:
-                return tmp_arg_code
-            return None
+        def find_templates(signature, templates):
+            new_templates = {d_type['datatype']:templates[d_type['datatype']]\
+                    for d_type in signature\
+                    if 'datatype' in d_type and d_type['datatype'] in templates}
+            return new_templates
```

Comment:
Even though this is a local function, I think it's useful to add a one-line docstring to it.